File: examples/gimbal_control_example.cpp
#include "stdio.h"
#include <pthread.h>
#include <cstdlib>
#include <string>
using namespace std;

#include"payloadSdkInterface.h"

#if (CONTROL_METHOD == CONTROL_UART)
T_ConnInfo s_conn = {
    CONTROL_UART,
    payload_uart_port,
    payload_uart_baud
};
#else
T_ConnInfo s_conn = {
    CONTROL_UDP,
    udp_ip_target,
    udp_port_target
};
#endif


class gimbalSetModeExample : public PayloadSdkInterface
{
private:
	float currentPitch;
	float currentRoll;
	float currentYaw;
public:
    /**
	 * set gimbal stop and re-center the yaw angle.
     **/
	void setRecenterYaw();

	/**
	 * set gimbal stop and re-center the pitch angle.
	 **/
	void setRecenterPitch();
	
	/**
	 * Update all current angles of the gimbal.
	 * @param currentPitch: Current pitch angle
	 * @param currentRoll: Current roll angle
	 * @param currentYaw: Current yaw angle
	 **/
	void updateAllCurrentAngle(float currentPitch, float currentRoll, float currentYaw);
};

void 
gimbalSetModeExample::
setRecenterYaw()
{
	// Keep gimbal stop, delay in 1.5secs
	setGimbalSpeed(0, 0, 0, INPUT_SPEED);
	usleep(1500000);

	// Re-center the yaw angle
	setGimbalSpeed(this->currentPitch, this->currentRoll, 0, INPUT_ANGLE);

	// Wait until the yaw is nearly dead zone
	while(fabs(this->currentYaw) >= 0.2);
}

void 
gimbalSetModeExample::
setRecenterPitch()
{
	// Keep gimbal stop, delay in 1.5secs
	setGimbalSpeed(0, 0, 0, INPUT_SPEED);
	usleep(1500000);

	// Re-center the pitch angle
	setGimbalSpeed(0, this->currentRoll, this->currentYaw, INPUT_ANGLE);

	// Wait until the pitch is nearly dead zone
	while(fabs(this->currentPitch) >= 0.2);
}

void 
gimbalSetModeExample::
updateAllCurrentAngle(float currentPitch, float currentRoll, float currentYaw)
{
	this->currentPitch = currentPitch;
	this->currentRoll = currentRoll;	
	this->currentYaw = currentYaw;
}

gimbalSetModeExample *my_payload = nullptr;
void quit_handler(int sig);
static void onPayloadStatusChanged(int event, double* param);
static void onPayloadParamChanged(int event, char* param_char, double* param_double);

int main(int argc, char *argv[]){
	printf("Starting Set gimbal mode example...\n");
	signal(SIGINT,quit_handler);

	// create payloadsdk object
	my_payload = new gimbalSetModeExample();

	// init payload
	my_payload->sdkInitConnection();
	printf("Waiting for payload signal! \n");

	// register callback function
	my_payload->regPayloadStatusChanged(onPayloadStatusChanged);

	// register callback function for parameter changes
	my_payload->regPayloadParamChanged(onPayloadParamChanged);

	// check connection
	my_payload->checkPayloadConnection();
	
	// set FOLLOW mode
	printf("Gimbal set mode FOLLOW, delay in 3 secs \n");
	my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_GIMBAL_MODE, PAYLOAD_CAMERA_GIMBAL_MODE_FOLLOW, PARAM_TYPE_UINT32);
	usleep(3000000);

	printf("Move gimbal angular velocity: pitch = -20 deg/s, yaw = 50 deg/s, delay in 3 secs \n");
	my_payload->setGimbalSpeed(-20, 0, 50, INPUT_SPEED);
	usleep(3000000);

	printf("Gimbal set re-center YAW ------ \n");
	my_payload->setRecenterYaw();

	printf("Gimbal set re-center PITCH ------ \n");
	my_payload->setRecenterPitch();

	printf("Move gimbal angular position: pitch = -40 deg, yaw = 90 deg, delay in 3 secs \n");
	my_payload->setGimbalSpeed(-40, 0, 90, INPUT_ANGLE);
	usleep(3000000);

	printf("Gimbal set re-center PITCH ------ \n");
	my_payload->setRecenterPitch();

	printf("Gimbal set re-center YAW ------ \n");
	my_payload->setRecenterYaw();

	printf("Delay in 3 secs before changing to LOCK MODE\n");
	usleep(3000000);	

	// set LOCK mode
	printf("Gimbal set mode LOCK, delay in 3 secs \n");	
	my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_GIMBAL_MODE, PAYLOAD_CAMERA_GIMBAL_MODE_LOCK, PARAM_TYPE_UINT32);
	usleep(3000000);

	printf("Move gimbal angular velocity: pitch = 20 deg/s, yaw = -50 deg/s, delay in 3 secs \n");
	my_payload->setGimbalSpeed(20, 0, -50, INPUT_SPEED);
	usleep(3000000);	

	printf("Gimbal set re-center PITCH ------ \n");
	my_payload->setRecenterPitch();

	printf("Gimbal set re-center YAW ------ \n");
	my_payload->setRecenterYaw();

	printf("Move gimbal angular position: pitch = 40 deg, yaw = -90 deg, delay in 3 secs \n");
	my_payload->setGimbalSpeed(40, 0, -90, INPUT_ANGLE);
	usleep(3000000);

	printf("Gimbal set re-center PITCH ------ \n");
	my_payload->setRecenterPitch();

	printf("Gimbal set re-center YAW ------ \n");
	my_payload->setRecenterYaw();
	usleep(3000000);	

	printf("Delay in 3 secs before changing to FOLLOW MODE\n");
	usleep(3000000);	

	// set FOLLOW mode
	printf("Gimbal set mode FOLLOW, delay in 3 secs \n");
	my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_GIMBAL_MODE, PAYLOAD_CAMERA_GIMBAL_MODE_FOLLOW, PARAM_TYPE_UINT32);
	usleep(3000000);

	printf("Move gimbal angular velocity: pitch = -20 deg/s, yaw = 50 deg/s, delay in 3 secs \n");
	my_payload->setGimbalSpeed(-20, 0, 50, INPUT_SPEED);
	usleep(3000000);

	printf("Gimbal set re-center YAW ------ \n");
	my_payload->setRecenterYaw();

	printf("Gimbal set re-center PITCH ------ \n");
	my_payload->setRecenterPitch();

	printf("Move gimbal angular position: pitch = -40 deg, yaw = 90 deg, delay in 3 secs \n");
	my_payload->setGimbalSpeed(-40, 0, 90, INPUT_ANGLE);
	usleep(3000000);

	printf("Gimbal set re-center PITCH ------ \n");
	my_payload->setRecenterPitch();

	printf("Gimbal set re-center YAW ------ \n");
	my_payload->setRecenterYaw();

	printf("Delay in 3 secs before changing to MAPPING MODE\n");
	usleep(3000000);	

	// set MAPPING mode
	printf("Gimbal set mode MAPPING, delay in 5 secs \n");	
	my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_GIMBAL_MODE, PAYLOAD_CAMERA_GIMBAL_MODE_MAPPING, PARAM_TYPE_UINT32);
	usleep(5000000);

	// set RESET mode
	printf("Gimbal set mode RESET, delay in 5 secs \n");	
	my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_GIMBAL_MODE, PAYLOAD_CAMERA_GIMBAL_MODE_RESET, PARAM_TYPE_UINT32);
	usleep(5000000);	

	// set OFF mode
	printf("Gimbal set mode OFF, delay in 5 secs \n");	
	my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_GIMBAL_MODE, PAYLOAD_CAMERA_GIMBAL_MODE_OFF, PARAM_TYPE_UINT32);	
	usleep(5000000);	
		
	// close payload interface
	try {
		my_payload->sdkQuit();
	}
	catch (int error){}

	return 0;
}

void quit_handler( int sig ){
    printf("\n");
    printf("TERMINATING AT USER REQUEST \n");
    printf("\n");

    // close payload interface
    try {
       my_payload->sdkQuit();
    }
    catch (int error){}

    // end program here
    exit(0);
}

static void onPayloadStatusChanged(int event, double* param)
{
	switch(event)
	{
		case PAYLOAD_GB_ATTITUDE:
		{
			float pitch = param[0];
			float roll = param[1];
			float yaw = param[2];
//			printf("Gimbal attitude: pitch = %.2f, roll = %.2f, yaw = %.2f\n", pitch, roll, yaw);
			// Update current angles
			my_payload->updateAllCurrentAngle(pitch, roll, yaw);
			break;
		}
		default: break;
	}
}

void onPayloadParamChanged(int event, char* param_char, double* param_double)
{
	switch(event){
	case PAYLOAD_GB_ATTITUDE:
	{
//		printf("GB MODE: %s\n", param_char);
		break;
	}
	default: break;
	}
}

File: examples/gimbal_control_example.py
#!/usr/bin/env python3
import sys
import os

# Add the libs directory to the path
sys.path.append(os.path.join(os.path.dirname(__file__), '..', 'libs'))

# Import config first to setup environment automatically
from config import config

import time
import signal
import sys
from pymavlink import mavutil
from payload_sdk import PayloadSdkInterface, payload_status_event_t, input_mode_t
from payload_define import *

my_payload = None

# Gimbal control class
class GimbalSetModeExample(PayloadSdkInterface):
    def __init__(self):
        super().__init__()
        self.currentPitch: float = 0.0
        self.currentRoll:  float = 0.0
        self.currentYaw:   float = 0.0

    # Set gimbal stop and re-center the yaw angle.
    def setRecenterYaw(self) -> None:
        # Keep gimbal stopped, delay 1.5s
        self.setGimbalSpeed(0, 0, 0, input_mode_t.INPUT_SPEED)
        time.sleep(1.5)

        # Re-center the yaw angle
        self.setGimbalSpeed(self.currentPitch, self.currentRoll, 0, input_mode_t.INPUT_ANGLE)

        # Wait until yaw nearly zero (busy-wait like C++)
        while abs(self.currentYaw) >= 0.2:
            pass  

    # Set gimbal stop and re-center the pitch angle.
    def setRecenterPitch(self) -> None:
        # Keep gimbal stopped, delay 1.5s
        self.setGimbalSpeed(0, 0, 0, input_mode_t.INPUT_SPEED)
        time.sleep(1.5)

        # Re-center the pitch angle
        self.setGimbalSpeed(0, self.currentRoll, self.currentYaw, input_mode_t.INPUT_ANGLE)

        # Wait until pitch nearly zero (busy-wait like C++)
        while abs(self.currentPitch) >= 0.2:
            pass  

    # Update all current angles of the gimbal (no print to match C++)
    def updateAllCurrentAngle(self, currentPitch: float, currentRoll: float, currentYaw: float) -> None:
        self.currentPitch = currentPitch
        self.currentRoll = currentRoll
        self.currentYaw = currentYaw

# Signal handler for quitting
def quit_handler(sig, frame):
    global my_payload
    print("\nTERMINATING AT USER REQUEST\n")
    try:
        my_payload.sdkQuit()
    except Exception:
        pass
    sys.exit(0)

# Callback function for payload status changes
def onPayloadStatusChanged(event: int, param: list):
    global my_payload
    if payload_status_event_t(event) == payload_status_event_t.PAYLOAD_GB_ATTITUDE:
        pitch = param[0]
        roll  = param[1]
        yaw   = param[2]
        my_payload.updateAllCurrentAngle(pitch, roll, yaw)

# Callback function for payload param changes   
def onPayloadParamChanged(event: int, param_char: str, param: list):
    if payload_status_event_t(event) == payload_status_event_t.PAYLOAD_GB_ATTITUDE:
        pass  

def main():
    global my_payload
    print("Starting Set gimbal mode example...")
    signal.signal(signal.SIGINT, quit_handler)

    # Create payloadsdk object
    my_payload = GimbalSetModeExample()

    # Init payload
    my_payload.sdkInitConnection()
    print("Waiting for payload signal!")

    # Register callback function
    my_payload.regPayloadStatusChanged(onPayloadStatusChanged)
    my_payload.regPayloadParamChanged(onPayloadParamChanged)

    # Check connection
    my_payload.checkPayloadConnection()

    # Set FOLLOW mode
    print("Gimbal set mode FOLLOW, delay in 3 secs")
    my_payload.setPayloadCameraParam(PAYLOAD_CAMERA_GIMBAL_MODE, payload_camera_gimbal_mode.PAYLOAD_CAMERA_GIMBAL_MODE_FOLLOW, mavutil.mavlink.MAV_PARAM_TYPE_UINT32)
    time.sleep(3.0)

    print("Move gimbal angular velocity: pitch = -20 deg/s, yaw = 50 deg/s, delay in 3 secs")
    my_payload.setGimbalSpeed(-20, 0, 50, input_mode_t.INPUT_SPEED)
    time.sleep(3.0)

    print("Gimbal set re-center YAW ------")
    my_payload.setRecenterYaw()

    print("Gimbal set re-center PITCH ------")
    my_payload.setRecenterPitch()

    print("Move gimbal angular position: pitch = -40 deg, yaw = 90 deg, delay in 3 secs")
    my_payload.setGimbalSpeed(-40, 0, 90, input_mode_t.INPUT_ANGLE)
    time.sleep(3.0)

    print("Gimbal set re-center PITCH ------")
    my_payload.setRecenterPitch()

    print("Gimbal set re-center YAW ------")
    my_payload.setRecenterYaw()

    print("Delay in 3 secs before changing to LOCK MODE")
    time.sleep(3.0)

    # Set LOCK mode
    print("Gimbal set mode LOCK, delay in 3 secs")
    my_payload.setPayloadCameraParam(PAYLOAD_CAMERA_GIMBAL_MODE, payload_camera_gimbal_mode.PAYLOAD_CAMERA_GIMBAL_MODE_LOCK, mavutil.mavlink.MAV_PARAM_TYPE_UINT32)
    time.sleep(3.0)

    print("Move gimbal angular velocity: pitch = 20 deg/s, yaw = -50 deg/s, delay in 3 secs")
    my_payload.setGimbalSpeed(20, 0, -50, input_mode_t.INPUT_SPEED)
    time.sleep(3.0)

    print("Gimbal set re-center PITCH ------")
    my_payload.setRecenterPitch()

    print("Gimbal set re-center YAW ------")
    my_payload.setRecenterYaw()

    print("Move gimbal angular position: pitch = 40 deg, yaw = -90 deg, delay in 3 secs")
    my_payload.setGimbalSpeed(40, 0, -90, input_mode_t.INPUT_ANGLE)
    time.sleep(3.0)

    print("Gimbal set re-center PITCH ------")
    my_payload.setRecenterPitch()

    print("Gimbal set re-center YAW ------")
    my_payload.setRecenterYaw()

    print("Delay in 3 secs before changing to FOLLOW MODE")
    time.sleep(3.0)

    # Set FOLLOW mode again
    print("Gimbal set mode FOLLOW, delay in 3 secs")
    my_payload.setPayloadCameraParam(PAYLOAD_CAMERA_GIMBAL_MODE, payload_camera_gimbal_mode.PAYLOAD_CAMERA_GIMBAL_MODE_FOLLOW, mavutil.mavlink.MAV_PARAM_TYPE_UINT32)
    time.sleep(3.0)

    print("Move gimbal angular velocity: pitch = -20 deg/s, yaw = 50 deg/s, delay in 3 secs")
    my_payload.setGimbalSpeed(-20, 0, 50, input_mode_t.INPUT_SPEED)
    time.sleep(3.0)

    print("Gimbal set re-center YAW ------")
    my_payload.setRecenterYaw()

    print("Gimbal set re-center PITCH ------")
    my_payload.setRecenterPitch()

    print("Move gimbal angular position: pitch = -40 deg, yaw = 90 deg, delay in 3 secs")
    my_payload.setGimbalSpeed(-40, 0, 90, input_mode_t.INPUT_ANGLE)
    time.sleep(3.0)

    print("Gimbal set re-center PITCH ------")
    my_payload.setRecenterPitch()

    print("Gimbal set re-center YAW ------")
    my_payload.setRecenterYaw()

    print("Delay in 3 secs before changing to MAPPING MODE")
    time.sleep(3.0)

    # Set MAPPING mode
    print("Gimbal set mode MAPPING, delay in 5 secs")
    my_payload.setPayloadCameraParam(PAYLOAD_CAMERA_GIMBAL_MODE, payload_camera_gimbal_mode.PAYLOAD_CAMERA_GIMBAL_MODE_MAPPING, mavutil.mavlink.MAV_PARAM_TYPE_UINT32)
    time.sleep(5.0)

    # Set OFF mode
    print("Gimbal set mode OFF, delay in 5 secs")
    my_payload.setPayloadCameraParam(PAYLOAD_CAMERA_GIMBAL_MODE, payload_camera_gimbal_mode.PAYLOAD_CAMERA_GIMBAL_MODE_OFF, mavutil.mavlink.MAV_PARAM_TYPE_UINT32)
    time.sleep(5.0)

    # Set RESET mode
    print("Gimbal set mode RESET, delay in 5 secs")
    my_payload.setPayloadCameraParam(PAYLOAD_CAMERA_GIMBAL_MODE, payload_camera_gimbal_mode.PAYLOAD_CAMERA_GIMBAL_MODE_RESET, mavutil.mavlink.MAV_PARAM_TYPE_UINT32)
    time.sleep(5.0)

    # Close payload interface
    try:
        my_payload.sdkQuit()
    except Exception:
        pass

if __name__ == "__main__":
    main()